### PR TITLE
DOCSP-43200: Get started with Django

### PR DESCRIPTION
**Pull request by norareidy**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-node/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-43200
Staging - https://deploy-preview-132--docs-pymongo.netlify.app/django-get-started/

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [ ] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: snooty.toml [hunk 1]
```diff
@@ -25,12 +26,14 @@ sharedinclude_root = "https://raw.githubusercontent.com/10gen/docs-shared/main/"
 driver-short = "PyMongo"
 driver-long = "PyMongo, the MongoDB synchronous Python driver,"
 driver-async = "PyMongo Async"
+django-odm = "MongoDB Backend for Django"
```
Comment (aclark4life) [inline snooty.toml:29]: `django-odm` == `django-mongodb-backend` ?

File: source/django-get-started.txt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+.. facet::
+   :name: genre
+   :values: tutorial
+ 
+.. meta::
+   :description: Learn how to create an app to connect to a MongoDB deployment by using Django MongoDB.
```
Comment (aclark4life) [inline source/django-get-started.txt:18]: `s/Django MongoDB/Django MongoDB Backend/`

File: source/index.txt [hunk 3]
```diff
@@ -13,6 +13,7 @@ MongoDB {+driver-short+} Documentation
 .. toctree::
 
    Get Started </get-started>
+   Get Started with Django MongoDB </django-get-started>
```
Comment (aclark4life) [inline source/index.txt:16]: `s/Django MongoDB/Django MongoDB Backend/`

**Review by aclark4life — CHANGES_REQUESTED** [review]
LGTM pending various small changes, thanks @norareidy

File: snooty.toml (hunk 1 above)
Comment (norareidy) [reply snooty.toml:29]: Yes! The name of the source constant isn't visible - it renders as "MongoDB Backend for Django" and we just use it for writing simplicity/consistency.

File: source/django-get-started/django-connect-mongodb.txt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+               "HOST": "<connection string URI>",
+            },
+         }
+
+      Replace the ``<username>`` and ``<password>`` placeholders with your
+      Atlas database user's username and password. Then, replace the
+      ``<connection string URI>`` placeholder with the connection string
+      that you copied from the :ref:`django-get-started-connection-string`
+      step of this guide. This configures your Django app to connect to
+      your Atlas deployment and access the ``sample_mflix`` sample database.
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started/django-connect-mongodb.txt:81]: The user's URI might look different, since we have them create their own - cluster, domain, etc might differ. So this would require a few different placeholders. I could just have them paste the full URI but they need to add the /sample_mflix part

[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started/django-connect-mongodb.txt:81]: I think keeping the full DATABASES setting is easier in this case

File: source/django-get-started/django-connection-string.txt [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+==========================
+Create a Connection String
+==========================
+
+You can connect to your MongoDB deployment by providing a
+**connection URI**, also called a *connection string*, which
+instructs the driver on how to connect to a MongoDB deployment
+and how to behave while connected.
+
+The connection string includes the hostname or IP address and
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started/django-connection-string.txt:12]: The URI that atlas generates doesn't include the database name, so they'd need to specify that separately (either with the NAME setting or editing the connection string)

**Review by aclark4life — APPROVED** [review]
LGTM, thanks

File: source/django-get-started/django-create-app.txt [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+   .. step:: Create models for movie, awards, and viewer data
+
+      Open the ``models.py`` file in the ``sample_mflix`` directory and replace 
+      its contents with the following code:
+
+      .. code-block:: python
+
+         from django.db import models
+         from django_mongodb_backend.fields import EmbeddedModelField, ArrayField
```
Comment (norareidy) [inline source/django-get-started/django-create-app.txt:49]: Note: the EmbeddedModelField import and the `awards` Movie field below will cause errors until [this PR](https://github.com/mongodb-labs/django-mongodb-backend/pull/151) is merged

**Review by Jibola — APPROVED** [review]
LGTM!!

File: source/django-get-started/django-create-admin.txt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+.. facet::
+   :name: genre
+   :values: tutorial
+
+.. meta::
+   :keywords: app, odm, code example
+
+You can create a Django admin site to edit your application's 
+data from a web interface. To learn more about the Django admin
+site and its features, see `The Django admin site <https://docs.djangoproject.com/en/5.1/ref/contrib/admin/>`__
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started/django-create-admin.txt:16]: Updated to use the django-version-number constant

File: source/django-get-started/django-install.txt [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+         pip install django-mongodb-backend
+
+      This command also installs the following dependencies:
+
+      - PyMongo version {+version-number+} and its dependencies
+      - Django version {+django-version-number+} and its dependencies
+
+      .. important::
+
+         If you have already installed PyMongo and Django in your
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started/django-install.txt:73]: Ah okay, deleted

File: source/django-get-started/django-write-data.txt [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+      .. code-block:: bash
+
+         python3 manage.py shell
+
+   .. step:: Import the required classes and modules
+
+      From your Python shell, run the following code to import
+      your models and the modules for creating a ``datetime`` object:
+
+      .. code-block:: python
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started/django-write-data.txt:36]: Our style guide advises against adding $ to bash examples, since that causes issues if the user copies and pastes directly into their terminal. I left out >>> for the same reason

File: source/django-get-started/django-connect-mongodb.txt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+               "PASSWORD": "<password>",
+               "HOST": "<connection string URI>",
+            },
+         }
+
+      Replace the ``<username>`` and ``<password>`` placeholders with your
+      Atlas database user's username and password. Then, replace the
+      ``<connection string URI>`` placeholder with the connection string
+      that you copied from the :ref:`django-get-started-connection-string`
+      step of this guide.
```
Comment (R-shubham) [inline source/django-get-started/django-connect-mongodb.txt]: The connection string that user copies will have username and password by default. If they are using this approach then "HOST" will expect connection string in this format: "mongodb+srv://cluster0.example.mongodb.net".  @Jibola / @aclark4life can you confirm?

File: snooty.toml [hunk 11]
```diff
@@ -25,12 +25,14 @@ sharedinclude_root = "https://raw.githubusercontent.com/10gen/docs-shared/main/"
 driver-short = "PyMongo"
 driver-long = "PyMongo, the MongoDB synchronous Python driver,"
 driver-async = "PyMongo Async"
+django-odm = "Django MongoDB"
```
Comment (aclark4life) [inline snooty.toml]: `s/Django MongoDB/Django MongoDB Backend/`?

File: source/django-get-started/django-connect-mongodb.txt (hunk 10 above)
Comment (aclark4life) [reply source/django-get-started/django-connect-mongodb.txt]: There are at least two things going on here:

- `DATABASES["default"]["HOST"]` can accept a prefix of `mongodb+srv://` but does not require one.
- Connection string is terminology only used in the context of `parse_uri` and in that context, the MONGODB_URI from Atlas can be "cut and pasted" with the exception of the `NAME` key which is not present in the Atlas MONGODB_URI but must be present in the `parse_uri` connection string.

To put it another way:

- `DATABASES["default"]["NAME"]` must be defined. -or-
- MONGODB_URI must include `/mydatabase` e.g. `mongodb+srv://cluster0.example.mongodb.net/mydatabase`.

Again, if it's easier to steer clear of `parse_uri` for clarity's sake that's OK.

File: source/django-get-started/django-connect-mongodb.txt [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      your Atlas deployment and access the ``sample_mflix`` sample database.
+
+   .. step:: Start the server
+
+      To verify that you installed {+django-odm+} and correctly configured
+      your project, run the following command from your project root:
+
+      .. code-block:: bash
+
+         python3 manage.py runserver
```
Comment (timgraham) [inline source/django-get-started/django-connect-mongodb.txt:90]: Maybe this is consistent with the rest of your documentation and you want to leave it alone, but if you are working in a Python 3 virtual environment, you can write `python` without the 3. Coming from Django's docs, it looks odd to include that 3 on all commands.

File: source/django-get-started/django-connect-mongodb.txt [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+                urls.py
+                asgi.py
+                wsgi.py
+
+   .. step:: Update your database settings
+
+      Open your ``settings.py`` file and navigate to the ``DATABASES`` setting.
+      Replace this setting with the following code:
+
+      .. code-block:: bash
```
Comment (timgraham) [inline source/django-get-started/django-connect-mongodb.txt:64]: python

File: source/django-get-started/django-connection-string.txt [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+      Paste this connection string into a file in your preferred text editor
+      and replace the ``<username>`` and ``<password>`` placeholders with
+      your database user's username and password.
+      
+      Save this file to a safe location for use in the next step.
+
+After completing these steps, you have a connection string that
+contains your database username and password.
+
```
Comment (timgraham) [inline source/django-get-started/django-connection-string.txt:56]: You may want to delete the stray blank line(s) at the end of some documents. ;-)

File: source/django-get-started/django-create-app.txt [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+
+         class Movie(models.Model):
+             title = models.CharField(max_length=200)
+             plot = models.TextField(blank=True)
+             runtime = models.IntegerField(default=0)
+             released = models.DateTimeField("release date", null=True, blank=True)
+             awards = EmbeddedModelField(Award)
+             genres = ArrayField(models.CharField(max_length=100), null=True, blank=True)
+             objects = MongoManager()
```
Comment (timgraham) [inline source/django-get-started/django-create-app.txt:58]: MongoManager isn't imported, but it's not required for this example.

File: source/django-get-started/django-create-app.txt [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+             path("admin/", admin.site.urls),
+         ]
+
+   .. step:: Create templates to format your data
+
+      In your ``sample_mflix`` directory, create a subdirectory called
+      ``templates``. Then, create a file called ``recent_movies.html``
+      and paste the following code:
+      
+      .. code-block:: python
```
Comment (timgraham) [inline source/django-get-started/django-create-app.txt:157]: html+django

File: source/django-get-started/django-create-app.txt [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+            </ul>
+         </body>
+         </html>
+      
+      This template formats the movie data requested by the ``recent_movies`` view.
+
+      Create another file in the ``sample_mflix/templates`` directory called
+      ``viewers_list.html`` and paste the following code:
+
+      .. code-block:: python
```
Comment (timgraham) [inline source/django-get-started/django-create-app.txt:186]: html+django

File: source/django-get-started/django-create-app.txt [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+
+   .. step:: Create migrations for your new models
+
+      From your project root, run the following command to create
+      migrations for the ``Movie``, ``Award``, and ``Viewer`` models and apply
+      the changes to the database:
+      
+      .. code-block:: bash
+
+         python3 manage.py makemigrations sample_mflix
```
Comment (timgraham) [inline source/django-get-started/django-create-app.txt:248]: sample_mflix is an existing database, so the models should have an inner `Meta` class with [`managed = False`](https://docs.djangoproject.com/en/dev/ref/models/options/#managed) so that Django doesn't try to create the collections. Otherwise, I believe MongoDB will error with "collection already exists) at https://github.com/mongodb-labs/django-mongodb-backend/blob/eae01a45820de3b3a0efab4c1e85faa96335b4ad/django_mongodb_backend/schema.py#L25.

File: source/django-get-started/django-create-app.txt [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+         python3 manage.py migrate
```
Comment (timgraham) [inline source/django-get-started/django-create-app.txt:249]: sample_mflix is an existing database, so the models should have an inner `Meta` class with [`managed = False`](https://docs.djangoproject.com/en/dev/ref/models/options/#managed) so that Django doesn't try to create (and otherwise modify) the collections. Otherwise, I believe MongoDB will error with "collection already exists) [here](https://github.com/mongodb-labs/django-mongodb-backend/blob/eae01a45820de3b3a0efab4c1e85faa96335b4ad/django_mongodb_backend/schema.py#L25.) when running migrate...

File: source/django-get-started/django-install.txt [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+      the Django integration:
+
+      .. code-block:: bash
+
+         pip install django-mongodb-backend
+
+      This command also installs the following dependencies:
+
+      - PyMongo version {+version-number+} and its dependencies
+      - Django version {+django-version-number+} and its dependencies
```
Comment (timgraham) [inline source/django-get-started/django-install.txt:66]: More precisely, this will be "The latest Django 5.0.x version...'

File: source/django-get-started/django-query-data.txt [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+      .. code-block:: python
+
+         from sample_mflix.models import Movie, Viewer
+         
+         Viewer.objects.filter(email="jason_momoa@gameofthron.es").first()
+
+      This code returns the name of the matching user:
+
+      .. code-block:: bash
```
Comment (timgraham) [inline source/django-get-started/django-query-data.txt:40]: Technically the formatter is "pycon" but since you don't format your console sessions normally, I'm not sure it matters.

File: source/django-get-started/django-write-data.txt [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+      Run the following code to create an ``Movie`` object that
+      stores data about a movie titled ``"Minari"``, including
+      its awards in an ``Award`` object:
+
+      .. code-block:: python
+         
+         movie_awards = Award(wins=122, nominations=245, text="Won 1 Oscar")
+         movie = Movie(title="Minari", plot=
+         "A Korean-American family moves to an Arkansas farm in search of their own American Dream",
+         runtime=217, released=timezone.make_aware(datetime(2020, 1, 26, 0, 0)),
```
Comment (timgraham) [inline source/django-get-started/django-write-data.txt:53]: `0, 0` is optional (default values)

File: source/django-get-started/django-write-data.txt [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+   .. step:: Insert a movie into the database
+
+      Run the following code to create an ``Movie`` object that
+      stores data about a movie titled ``"Minari"``, including
+      its awards in an ``Award`` object:
+
+      .. code-block:: python
+         
+         movie_awards = Award(wins=122, nominations=245, text="Won 1 Oscar")
+         movie = Movie(title="Minari", plot=
```
Comment (timgraham) [inline source/django-get-started/django-write-data.txt:51]: If you like, it would be more readable to format these using hanging indent, e.g. 
```
movie = Movie(
    title="Minari",
    plot="...",
    ...
)
```
Also, you can use `movie = Movie.objects.create(...)`if you want to avoid the separate `save()` step. That's the usual pattern unless you need to do something between initializing the model and saving it.

File: source/django-get-started/django-connect-mongodb.txt (hunk 12 above)
Comment (aclark4life) [reply source/django-get-started/django-connect-mongodb.txt:90]: I was going to raise this too, but would defer to MongoDB-wide conventions. If there are no such conventions, then I prefer `python` too.

File: source/django-get-started/django-create-app.txt (hunk 19 above)
Comment (aclark4life) [reply source/django-get-started/django-create-app.txt:249]: @timgraham As well as set `db_name` correct?

Comment (timgraham) [reply source/django-get-started/django-create-app.txt:249]: Yes, `db_table` is already set.

File: source/django-get-started.txt [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+   Create a Deployment </django-get-started/django-create-deployment/>
+   Create a Connection String </django-get-started/django-connection-string/>
+   Configure a MongoDB Connection </django-get-started/django-connect-mongodb/>
+   Create an Application </django-get-started/django-create-app/>
+   Write Data to MongoDB </django-get-started/django-write-data/>
+   Query MongoDB Data </django-get-started/django-query-data/>
+   Create an Admin Site </django-get-started/django-create-admin/>
+   Next Steps </django-get-started/django-next-steps/>
+
+{+django-odm+} is a PyMongo integration that allows you to use the
```
[the remark this answers is not in the document]
Comment (norareidy) [reply source/django-get-started.txt:33]: Updated! Do you have a preference between "Django MongoDB Backend" and "MongoDB Backend for Django"?

[the remark this answers is not in the document]
Comment (timgraham) [reply source/django-get-started.txt:33]: The other backends I've written use the latter phrasing, however, since we had to rename our package from django-mongondb to django-mongodb-backend, we can use the former.

File: source/django-get-started/django-query-data.txt (hunk 21 above)
Comment (norareidy) [reply source/django-get-started/django-query-data.txt:40]: pycon is not an option for the .. code-block:: directive unfortunately

**Review by gmiller-mdb — APPROVED** [review]
LGTM!

**Comment by norareidy** [comment]
Moved to the new docs-django PR: https://github.com/mongodb/docs-django/pull/10

**Review by R-shubham — COMMENTED** [review]
LGTM
